src/cast_core/src/Loader.cpp

```diff
@@ -38,27 +38,66 @@ using index_type = std::pair<DocumentGenerator, std::optional<DocumentGenerator>
+    PhaseConfig(PhaseContext& context,
+                mongocxx::pool::entry& client,
+                uint thread,
+                uint totalThreads,
+                ActorId id)
         : database{(*client)[context["Database"].to<std::string>()]},
-          // The next line uses integer division. The Remainder is accounted for below.
-          numCollections{context["CollectionCount"].to<IntegerSpec>() /
-                         context["Threads"].to<IntegerSpec>()},
+          multipleThreadsPerCollection{
+              context["MultipleThreadsPerCollection"].maybe<bool>().value_or(false)},
+          // The next line uses integer division for non multithreaded configurations.
+          // The Remainder is accounted for below.
+          numCollections{multipleThreadsPerCollection
+                             ? 1
+                             : context["CollectionCount"].to<IntegerSpec>() /
+                                 context["Threads"].to<IntegerSpec>()},
           numDocuments{context["DocumentCount"].to<IntegerSpec>()},
           batchSize{context["BatchSize"].to<IntegerSpec>()},
           documentExpr{context["Document"].to<DocumentGenerator>(context, id)},
-          collectionOffset{numCollections * thread} {
-        auto& indexNodes = context["Indexes"];
-        for (auto [k, indexNode] : indexNodes) {
-            indexes.emplace_back(indexNode["keys"].to<DocumentGenerator>(context, id),
-                                 indexNode["options"].maybe<DocumentGenerator>(context, id));
-        }
-        if (thread == context["Threads"].to<int>() - 1) {
-            // Pick up any extra collections left over by the division
-            numCollections += context["CollectionCount"].to<uint>() % context["Threads"].to<uint>();
+          collectionOffset{multipleThreadsPerCollection
+                               ? thread % context["CollectionCount"].to<IntegerSpec>()
+                               : numCollections * thread} {
+        auto createIndexes = [&]() {
+            auto& indexNodes = context["Indexes"];
+            for (auto [k, indexNode] : indexNodes) {
+                indexes.emplace_back(indexNode["keys"].to<DocumentGenerator>(context, id),
+                                     indexNode["options"].maybe<DocumentGenerator>(context, id));
+            }
+        };
+
+        if (multipleThreadsPerCollection) {
+            if (context["Threads"]) {
+                BOOST_THROW_EXCEPTION(InvalidConfigurationException(
+                    "Phase Config 'Threads' parameter is not supported if "
+                    "'MultipleThreadsPerCollection' is true."));
+            }
+            uint collectionCount = context["CollectionCount"].to<IntegerSpec>();
+            if (totalThreads % collectionCount != 0) {
+                BOOST_THROW_EXCEPTION(InvalidConfigurationException(
+                    "'CollectionCount' must be an even divisor of 'Threads'."));
```

Comment:
consider including the configured `CollectionCount` and/or other yaml params for an easier time debugging.

Reply:
good idea.